RDS new facts modules

willthames opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
rds_instance_facts and rds_snapshot_facts

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel b07aa990c9) last updated 2017/02/24 14:05:54 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This PR reflects the completed module work as a precursor to new, mostly backward compatible `rds_snapshot` and `rds_instance` modules of #19772.

mattclay commented on lib/ansible/module_utils/rds.py:
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/python
```
Files in module_utils shouldn't have a shebang.

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

Etherdaemon commented:
shipit

wimnat commented on lib/ansible/module_utils/rds.py:
```diff
@@ -0,0 +1,135 @@
+            d['vpc_security_groups'] = ','.join(x['VpcSecurityGroupId'] for x in self.instance['VpcSecurityGroups'])
+        if self.instance.get("Endpoint"):
+            d['endpoint'] = self.instance["Endpoint"].get('Address', None)
+            d['port'] = self.instance["Endpoint"].get('Port', None)
```
Missing HostedZoneId here which is important if you want to set a route53 ALIAS record later

wimnat commented:
There is a lot of "facts" missing from these modules.

Here's what i would expect back following a facts call:

```
"db_instances": [
    {
        "allocated_storage": 100,
        "auto_minor_version_upgrade": true,
        "availability_zone": "ap-southeast-2a",
        "backup_retention_period": 7,
        "ca_certificate_identifier": "rds-ca-2015",
        "copy_tags_to_snapshot": true,
        "db_instance_arn": "arn:aws:rds:ap-southeast-2:602986065330:db:rds-stg-mysql-1",
        "db_instance_class": "db.m3.large",
        "db_instance_identifier": "rds-stg-mysql-1",
        "db_instance_port": 0,
        "db_instance_status": "available",
        "db_parameter_groups": [
            {
                "db_parameter_group_name": "qlpos-mysql5-7",
                "parameter_apply_status": "in-sync"
            }
        ],
        "db_security_groups": [],
        "db_subnet_group": {
            "db_subnet_group_description": "staging",
            "db_subnet_group_name": "staging",
            "subnet_group_status": "Complete",
            "subnets": [
                {
                    "subnet_availability_zone": {
                        "name": "ap-southeast-2a"
                    },
                    "subnet_identifier": "subnet-cc03a5bb",
                    "subnet_status": "Active"
                },
                {
                    "subnet_availability_zone": {
                        "name": "ap-southeast-2b"
                    },
                    "subnet_identifier": "subnet-f69a2293",
                    "subnet_status": "Active"
                }
            ],
            "vpc_id": "vpc-65cf6a00"
        },
        "dbi_resource_id": "db-BAWBHITP543RZ75AZPUSFEHZLY",
        "domain_memberships": [],
        "endpoint": {
            "address": "rds-stg-mysql-1.cuvcyf0og5ui.ap-southeast-2.rds.amazonaws.com",
            "hosted_zone_id": "Z32T0VRHXEXS0V",
            "port": 3306
        },
        "engine": "mysql",
        "engine_version": "5.7.16",
        "instance_create_time": "2017-03-20T00:32:13.826000+00:00",
        "latest_restorable_time": "2017-03-20T04:50:00+00:00",
        "license_model": "general-public-license",
        "master_username": "sa",
        "monitoring_interval": 0,
        "multi_az": true,
        "option_group_memberships": [
            {
                "option_group_name": "default:mysql-5-7",
                "status": "in-sync"
            }
        ],
        "pending_modified_values": {},
        "preferred_backup_window": "13:51-14:21",
        "preferred_maintenance_window": "sat:17:31-sat:18:01",
        "publicly_accessible": true,
        "read_replica_db_instance_identifiers": [],
        "secondary_availability_zone": "ap-southeast-2b",
        "storage_encrypted": false,
        "storage_type": "gp2",
        "vpc_security_groups": [
            {
                "status": "active",
                "vpc_security_group_id": "sg-73df4a14"
            }
        ]
    }
]
```

Also, tags is not returned with a `describe_db_instances` call so you can't do `self.instance.get("Tags")`.  You'll need to do a further `list_tags_for_resource` call for each ARN returned.

willthames commented:
@wimnat - that is a fair observation - I was attempting to keep rds_instance_facts relatively backward compatible with current rds module but the current rds module is itself fairly limited.

If backward compatibility is less desirable than completeness of results (and it probably is), using camel_dict_to_snake_dict on the results and working from there is a good starting point.

Thanks for the observation on tags - I'm sure I meant to do that, but clearly didn't

Commit:
Provide all returned detail as facts

wimnat commented:
No worries. I've thrown a PR your way. :)


On 20 Mar. 2017 16:09, "Will Thames" <notifications@github.com> wrote:

@wimnat <https://github.com/wimnat> - that is a fair observation - I was
attempting to keep rds_instance_facts relatively backward compatible with
current rds module but the current rds module is itself fairly limited.

If backward compatibility is less desirable than completeness of results
(and it probably is), using camel_dict_to_snake_dict on the results and
working from there is a good starting point.

Thanks for the observation on tags - I'm sure I meant to do that, but
clearly didn't

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ansible/ansible/pull/21884#issuecomment-287679758>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AHFORlmWyhmJfS9S4FMHN6wLKEbJSz4Qks5rngn9gaJpZM4MKz0d>
.

willthames commented:
Rebased to avoid the merge commit

Commit:
rds facts: Update to latest metadata standards

Commit:
rds_snapshot_facts - fix pep8 issues

Commit:
Tidy up flake8 violations in module_utils.rds

mjschultz commented on lib/ansible/module_utils/rds.py:
```diff
@@ -0,0 +1,105 @@
+class RDSDBInstance(object):
+    def __init__(self, dbinstance):
+        self.instance = dbinstance
+        if 'DBInstanceIdentifier' not in dbinstance:
```
This is equivalent to:
```
self.name = self.instance.get('DBInstanceIdentifier')
```
isn't it?

willthames replied:
Ummm, yes.

willthames replied:
Fixed now

Commit:
Fix unnecessary dict handling code

Remove unnecessary dict checking code that the `get` method
handles just fine.

wimnat commented:
Is it worthwhile renaming the module utils file to aws_rds.py? Just thinking the aws distinction may be useful, particularly if more AWS utils files are forthcoming

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot_facts.py:
```diff
@@ -0,0 +1,164 @@
+version_added: "2.4"
+short_description: obtain facts about one or more RDS snapshots
+description:
+  - obtain facts about one or more RDS snapshots
```
Are there caveats to what kind of snapshots this is supposed to find? For instance, I have Aurora (MySQL-Compatible) and MySQL snapshots. It returns the MySQL snapshots, but does not appear to find the Aurora instance snapshots. Is this expected behavior?

willthames replied:
@s-hertel inconsistent AWS APIs is the real answer. Looks like Aurora uses [describe_db_cluster_snapshots](http://boto3.readthedocs.io/en/latest/reference/services/rds.html#RDS.Client.describe_db_cluster_snapshots)

We should be able to transparently support that easily enough

s-hertel replied:
Do you want to document the limitations or wait to merge this till we support Aurora as well? That was my only feedback.

michael-dev2rights replied:
@s-hertel in the [new version of this PR](https://github.com/ansible/ansible/pull/26599) I added documentation of the lack of support for Aurora.  I think it would be obviously better to fix this.

ryansb commented on lib/ansible/module_utils/rds.py:
```diff
@@ -0,0 +1,102 @@
+            raise
+
+
+class RDSDBInstance(object):
```
This feels to me like it would be better as a function that would take a dbinstance and convert it to the facts dict, then a separate one to diff dictionaries rather than a class.

ryansb commented on lib/ansible/module_utils/rds.py:
```diff
@@ -0,0 +1,102 @@
+        return not self == other
+
+
+class RDSSnapshot(object):
```
Same - this class seems more like a snapshot -> facts dict function than its own class.

ryansb commented on lib/ansible/module_utils/rds.py:
```diff
@@ -0,0 +1,102 @@
+            raise
+
+
+def get_db_snapshot(conn, snapshotid):
```
These should be using the backoff tools, since I can see folks grabbing facts on large numbers of snapshots.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,158 @@
+    results = list()
+    while True:
+        try:
+            response = conn.describe_db_instances(Marker=marker, **params)
```
Instead of manually paginating, use boto3's [get_paginator](http://boto3.readthedocs.io/en/latest/reference/services/rds.html#RDS.Client.get_paginator) then you can iterate over that.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,158 @@
+
+    # Get tags for each response
+    for instance in results:
+        instance['tags'] = boto3_tag_list_to_ansible_dict(conn.list_tags_for_resource(ResourceName=instance['DBInstanceArn'])['TagList'])
```
This part should definitely have AWSRetry around it, since you're going to be hitting list_tags a *ton*.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_snapshot_facts.py:
```diff
@@ -0,0 +1,164 @@
+    results = list()
+    while True:
+        try:
+            response = conn.describe_db_snapshots(Marker=marker, **params)
```
Same - use `get_paginator`

michael-dev2rights commented:
@wimnat I am not sure about renaming this rds module.  If we do that then probably all of them should be renamed at the same time?

Following agreement on #ansible-aws I have split this PR into three new PRs.  

* rds module utils  https://github.com/ansible/ansible/pull/26594 
* rds_instance_facts https://github.com/ansible/ansible/pull/26598
* rds_snapshot_facts https://github.com/ansible/ansible/pull/26599

I believe that all comments in this PR have been taken into account.  @willthames if you think my new PR is good could you please close this one.

willthames commented:
Superseded as @michael-dev2rights notes above
